pyro/infer/mcmc/adaptation.py

```diff
@@ -194,25 +176,177 @@ def step(self, t, z, accept_prob):
     def adaptation_schedule(self):
         return self._adaptation_schedule
 
+    @property
+    def mass_matrix_adapter(self):
+        return self._mass_matrix_adapter
+
+    @mass_matrix_adapter.setter
+    def mass_matrix_adapter(self, value):
+        self._mass_matrix_adapter = value
+
+
+# this works for diagonal matrix `x`
+def _matmul(x, y):
+    return x.mul(y) if x.dim() == 1 else x.matmul(y)
+
+
+def _cholesky(x):
+    return x.sqrt() if x.dim() == 1 else x.cholesky()
+
+
+def _transpose(x):
+    return x if x.dim() == 1 else x.t()
+
+
+def _upper_triangular_inverse(x):
+    if x.dim() == 1:
+        return x.reciprocal()
+    else:
+        identity = torch.eye(x.size(-1), dtype=x.dtype, device=x.device)
+        return torch.triangular_solve(identity, x, upper=True)[0]
+
+
+class BlockMassMatrix:
+    """
```

Comment:
Do you foresee any changes to the interface? If so, we should mark this as experimental in the docstring.

Reply:
I think the interface is good. We might add more features to the constructor if needed. But I'll mark this as Experimental.